[clang] Allow devirtualisation of indirect calls to final virtual methods

ojhunt opened the pull request:
When -fstrict-vtable-pointers is set we can devirtualise calls to
virtual functions when called indirectly through a separate function
that does not locally know the exact type it is operating on.

This only permits the optimization for regular methods, not any kind
of constructor or destructor.

ojhunt committed:
[clang] Allow devirtualisation of indirect calls to final virtual methods

When -fstrict-vtable-pointers is set we can devirtualise calls to
virtual functions when called indirectly through a separate function
that does not locally know the exact type it is operating on.

This only permits the optimization for regular methods, not any kind
of constructor or destructor.

ChuanqiXu9 reviewed (commented):
LGTM generally, thanks.

ojhunt committed:
Ensure that the test doesn't doesn't vary according to platform mangling

ojhunt commented:
> BTW, have you tested it with bootstrapped build?

no, but honestly my success rate at simply getting a trivial bootstrap build to work has not been great :-/

ojhunt committed:
increasing test coverage

ojhunt committed:
Added the strict-vtable test results, and added tests

New tests for multiple and virtual inheritance to verify that
we still perform this adjustments even when  devirtualising the
calls

ChuanqiXu9 commented:
> @ChuanqiXu9 I'm seeing bootstrap failure on main :-/

Will it fail with unmodified fstrict-vtable-pointers? Or if it will still fail if we give up the complex cases for multiple vtables? You can look at my patch for it (limits the number of vtables)

ChuanqiXu9 commented on clang/test/CodeGenCXX/indirect-final-vcall-through-base.cpp:
```diff
@@ -0,0 +1,414 @@
+
+// STRICT-LABEL: i64 @_ZNK8Derived18getBatchEmmPl
+// STRICT-NOT: call
+// STRICT: ret i64
```
Yeah, I am using `// RUN: %clang_cc1 -triple=x86_64 -std=c++26 %s -emit-llvm -O3 -fstrict-vtable-pointers -o - | FileCheck %s --check-prefix=STRICT`

ChuanqiXu9 replied:
I succeed with `// STRICT-NOT: call {{[^(]*}}%`

ojhunt replied:
What targets do you have configured? (I want to see if I can repro locally, and then make the tests robust against it)

ChuanqiXu9 replied:
Can we replace `// STRICT-NOT: call` with `// STRICT-NOT: call {{[^(]*}}%` I feel it makes it more clear that we're checking the dynamic calls are erased.

ChuanqiXu9 commented:
And also, when bootstrapping, have you met errors like:  `undefined symbol: clang::ImplicitConceptSpecializationDecl::~ImplicitConceptSpecializationDecl()`? There are other similar cases.

The common point of these decl is, they have a virtual destructor but they don't have any other virtual function. This is why my previous patch wrote: https://github.com/ChuanqiXu9/llvm-project/commit/40ceabea69d20e4c8b0e5415e4f6c0b37c6b863a#diff-9f23818ed51d0b117b5692129d0801721283d0f128a01cbc562353da0266d7adR1375-R1382

ChuanqiXu9 commented:
After I add the following, the bootstrap build and test succeeded for me:

```
    if (CGM.getCodeGenOpts().StrictVTablePointers && CXXThisValue) {
      const CXXRecordDecl *ThisRecordDecl = MD->getParent();
      bool IsPolymorphicObject = ThisRecordDecl->isPolymorphic();
      bool IsStructor = isa<CXXDestructorDecl, CXXConstructorDecl>(MD);
      bool IsFinal = ThisRecordDecl->isEffectivelyFinal();

      // A workaround for cases the base class defined virtual dtor while
      // the derived class don't offer it. Then the current implementation may
      // introduce undefined symbols as it introduce all symbols in the
      // vtable.
      bool BaseHasDtor = false;
      ThisRecordDecl->forallBases([&](const CXXRecordDecl *Base) -> bool {
        if (Base->getDestructor() && !Base->getDestructor()->isImplicit())
            BaseHasDtor = true;
  
        return true;
      });
      bool HasDtor = ThisRecordDecl->getDestructor() &&
          !ThisRecordDecl->getDestructor()->isImplicit();

      // We do not care about whether this is a virtual method, because even
      // if the current method is not virtual, it may be calling another method
      // that calls a virtual function.
      if (IsPolymorphicObject && !IsStructor && IsFinal && (!BaseHasDtor || HasDtor))
        EmitVTableAssumptionLoads(ThisRecordDecl, LoadCXXThisAddress());
    }
```

ChuanqiXu9 commented:
> Ok, bootstrap - assertions passes all tests for me, what is your bootstrap config?

LDFLAGS="-fuse-ld=lld -Wl,-q -Wl,-znow -Wl,-build-id=sha1" cmake -G Ninja  -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=... -DCMAKE_CXX_COMPILER=... -DCMAKE_C_FLAG="-fstack-clash-protection -fcf-protection -Wno-backend-plugin" -DCMAKE_CXX_FLAGS="-fstack-clash-protection -fcf-protection -Wno-backend-plugin -fstrict-vtable-pointers" -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;openmp" -DLLVM_ENABLE_RUNTIMES="compiler-rt" -DLLVM_LIBDIR_SUFFIX=64 -DCLANG_LIBDIR_SUFFIX=64 -DOPENMP_LIBDIR_SUFFIX=64 -DLLVM_TARGETS_TO_BUILD="X86;AArch64;BPF;WebAssembly" -DLLVM_ENABLE_LIBCXX=ON -DLLVM_STATIC_LINK_CXX_STDLIB=ON -DCOMPILER_RT_BUILD_LIBFUZZER=OFF  -DCOMPILER_RT_BUILD_SANITIZERS=OFF -DCOMPILER_RT_BUILD_XRAY=OFF -DCOMPILER_RT_BUILD_ORC=OFF -DLLVM_USE_LINKER=lld ../llvm

ChuanqiXu9 commented on clang/test/CodeGenCXX/indirect-final-vcall-through-base.cpp:
```diff
@@ -0,0 +1,414 @@
+// STRICT-LABEL: i64 @_ZN8Derived312callGetValueEv(
+// STRICT: [[TRUE_THIS:%.*]] = getelementptr inbounds nuw i8, ptr %this, i64 16
+// STRICT: [[INVARIANT_THIS:%.*]] = tail call ptr @llvm.strip.invariant.group.p0(ptr nonnull [[TRUE_THIS]])
+// STRICT: [[VALUE_PTR:%.*]] = getelementptr inbounds nuw i8, ptr [[INVARIANT_THIS]], i64 8
```
I am slightly confused about the intention for these checks. Why do we check there is a GEP?

ojhunt commented:
> LGTM generally, so what's the problem for now?

Review :D

ojhunt commented on clang/test/CodeGenCXX/indirect-final-vcall-through-base.cpp:
```diff
@@ -0,0 +1,414 @@
+// mangling and similar is consistent. Choosing x86_64 as that seems to be a
+// configured target for most build configurations
+// RUN: %clang_cc1 -triple=x86_64 -std=c++26 %s -emit-llvm -O3                          -o - | FileCheck %s
+// RUN: %clang_cc1 -triple=x86_64 -std=c++26 %s -emit-llvm -O3 -fstrict-vtable-pointers -o - | FileCheck %s --check-prefix=STRICT
```
👍
